src/core/Akka.Persistence/RecoveryPermitter.cs

```diff
@@ -0,0 +1,94 @@
+﻿using System;
+using System.Collections.Generic;
+using Akka.Actor;
+using Akka.Event;
+using Akka.Pattern;
+using Akka.Persistence.Internal;
+
+namespace Akka.Persistence
+{
+    internal class RequestRecoveryPermit
+    { }
+
+    internal class RecoveryPermitGranted
+    { }
+
+    internal class ReturnRecoveryPermit
+    { }
+
+    /// <summary>
+    /// When starting many persistent actors at the same time the journal its data store is protected 
+    /// from being overloaded by limiting number of recoveries that can be in progress at the same time.
+    /// </summary>
+    internal class RecoveryPermitter : UntypedActor
+    {
+        private readonly int maxPermits;
+        private readonly LinkedList<IActorRef> pending = new LinkedList<IActorRef>();
+        private readonly ILoggingAdapter Log = Context.GetLogger();
+        private int usedPermits;
+        private int maxPendingStats;
+
+        public static Props Props(int maxPermits) =>
+            Actor.Props.Create(() => new RecoveryPermitter(maxPermits));
+
+        public RecoveryPermitter(int maxPermits)
+        {
+            this.maxPermits = maxPermits;
+        }
+
+        protected override void OnReceive(object message)
+        {
+            if (message is RequestRecoveryPermit)
+            {
+                Context.Watch(Sender);
+                if (usedPermits >= maxPermits)
+                {
+                    if (pending.Count == 0)
+                        Log.Debug("Exceeded max-concurrent-recoveries [{0}]. First pending {1}", maxPermits, Sender);
+                    pending.AddLast(Sender);
+                    maxPendingStats = Math.Max(maxPendingStats, pending.Count);
+                }
+                else
+                {
+                    RecoveryPermitGranted(Sender);
+                }
+            }
+            else if (message is ReturnRecoveryPermit)
+            {
+                ReturnRecoveryPermit(Sender);
+            }
+            else if (message is Terminated terminated && !pending.Remove(terminated.ActorRef))
+            {
+                // pre-mature termination should be rare
+                ReturnRecoveryPermit(terminated.ActorRef);
+            }
+        }
+
+        private void ReturnRecoveryPermit(IActorRef actorRef)
+        {
+            usedPermits--;
+            Context.Unwatch(actorRef);
+
+            if (usedPermits < 0)
+                throw new IllegalStateException("Permits must not be negative");
+
+            if (pending.Count > 0)
+            {
+                var popRef = pending.Pop();
```

Comment:
[Seems like it](https://stackoverflow.com/a/35547316/465132), though I'm not 100% sure